server/store/middleware/swarm/swarm.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2016 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package response
+
+import (
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/pkg/event"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("store_swarm_interaction", announceSwarmInteraction)
+}
+
+// FailedSwarmInteraction represents an error that indicates that the
+// interaction of a peer with a swarm failed.
+type FailedSwarmInteraction string
+
+// Error satisfies the error interface for FailedSwarmInteraction.
+func (f FailedSwarmInteraction) Error() string { return string(f) }
+
+// announceSwarmInteraction provides a middleware that manages swarm
+// interactions for a peer based on the announce.
+func announceSwarmInteraction(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
+		if req.IPv4 != nil {
+			err = updatePeerStore(req, req.Peer4())
+		}
+		if req.IPv6 != nil {
+			err = updatePeerStore(req, req.Peer6())
+		}
+
+		if err != nil {
+			return FailedSwarmInteraction(err.Error())
+		}
+
+		return next(cfg, req, resp)
+	}
+}
+
+func updatePeerStore(req *chihaya.AnnounceRequest, peer chihaya.Peer) (err error) {
+	storage := store.MustGetStore()
+
+	switch req.Event {
+	case event.Completed:
+		err = storage.GraduateLeecher(req.InfoHash, peer)
+	case event.Stopped:
+		if req.Left > 0 {
+			err = storage.DeleteLeecher(req.InfoHash, peer)
+		} else {
+			err = storage.DeleteSeeder(req.InfoHash, peer)
+		}
+	default:
+		if req.Left > 0 {
+			err = storage.PutLeecher(req.InfoHash, peer)
+		} else {
+			err = storage.PutSeeder(req.InfoHash, peer)
+		}
+	}
+
+	return
```

Comment:
Yeah that looks pretty good, I would leave out the paused event though - we don't support the partial seeds extension to scrapes and I think we'd have to handle even more corner cases if we handle it.